components/constellation/constellation.rs

```diff
@@ -3380,6 +3274,113 @@ where
+
+            // The past is stored with older entries at the front.
+            // We reverse the iter so that newer entries are at the front and then
+            // skip _n_ entries and evict the remaining entries.
+            let mut pipelines_to_evict = session_history
+                .past
+                .iter()
+                .rev()
+                .map(|diff| diff.alive_old_pipeline())
+                .skip(history_length)
+                .filter_map(|maybe_pipeline| maybe_pipeline)
+                .collect::<Vec<_>>();
+
+            // The future is stored with oldest entries front, so we must
+            // reverse the iterator like we do for the `past`.
+            pipelines_to_evict.extend(
+                session_history
+                    .future
+                    .iter()
+                    .rev()
+                    .map(|diff| diff.alive_new_pipeline())
+                    .skip(history_length)
+                    .filter_map(|maybe_pipeline| maybe_pipeline),
+            );
+
+            pipelines_to_evict
+        };
+
+        let mut dead_pipelines = vec![];
+        for evicted_id in pipelines_to_evict {
+            let load_data = match self.pipelines.get(&evicted_id) {
+                Some(pipeline) => {
+                    let mut load_data = pipeline.load_data.clone();
+                    load_data.url = pipeline.url.clone();
+                    load_data
+                },
+                None => continue,
+            };
+
+            dead_pipelines.push((evicted_id, NeedsToReload::Yes(evicted_id, load_data)));
+            self.close_pipeline(
+                evicted_id,
+                DiscardBrowsingContext::No,
+                ExitPipelineMode::Normal,
+            );
+        }
+
+        let session_history = self.get_joint_session_history(top_level_browsing_context_id);
+
+        for (alive_id, dead) in dead_pipelines {
+            session_history.replace_reloader(NeedsToReload::No(alive_id), dead);
+        }
+    }
+
+    fn handle_activate_document_msg(&mut self, pipeline_id: PipelineId) {
+        debug!("Document ready to activate {}", pipeline_id);
+
+        // Find the pending change whose new pipeline id is pipeline_id.
+        let pending_index = self
+            .pending_changes
+            .iter()
+            .rposition(|change| change.new_pipeline_id == pipeline_id);
+
+        // If it is found, remove it from the pending changes, and make it
+        // the active document of its frame.
+        if let Some(pending_index) = pending_index {
+            let change = self.pending_changes.swap_remove(pending_index);
+            // Notify the parent (if there is one).
+            let parent_pipeline_id = match change.new_browsing_context_info {
+                // This will be a new browsing context.
+                Some(ref info) => info.parent_pipeline_id,
+                // This is an existing browsing context. TODO this is an unfortunate effect of
```

Comment:
I think this lookup makes sense here. Not sure what an alternative would be.

Reply:
Me neither, but it bothered me a little. I'll remove the TODO then.